scipy/sparse/csgraph/_disjoint_set.py

```diff
@@ -0,0 +1,126 @@
+"""
+Disjoint set data structure
+"""
+
+
+class DisjointSet:
+    """ Disjoint set data structure for incremental connectivity queries.
+
+    .. versionadded: 1.6.0
+
+    Attributes
+    ----------
+    n_nodes : int
+        The number of nodes in the set.
+    n_components : int
+        The number of components/subsets.
+
+    Methods
+    -------
+    union
+    find
+
+    Notes
+    -----
+    This class implements the disjoint set [1]_, also known as the *union-find*
+    data structure. The *find* method implements the *path halving*
+    variant. The *union* method implements the *union by size* variant.
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+
+    See Also
+    --------
+    connected_components : Analyze the connected components of a sparse graph.
+
+    Examples
+    --------
+    >>> from scipy.sparse.csgraph import DisjointSet
+
+    Initialize a disjoint set:
+
+    >>> dis = DisjointSet()
+
+    Merge some subsets:
+
+    >>> dis.union(1, 2)
+    True
+    >>> dis.union(3, 'x')
+    True
+    >>> dis.union('x', 'y')
+    True
+    >>> dis.union('y', 'y')
+    False
+
+    Find root nodes:
+
+    >>> dis[2]
+    1
+    >>> dis['y']
+    3
+
+    """
+    def __init__(self):
+        self.n_nodes = 0
+        self.n_components = 0
+        self._sizes = {}
+        self._parents = {}
+        self._indices = {}
+
+    def __getitem__(self, x):
+        """Find the root node of `x`.
+
+        Parameters
+        ----------
+        x : immutable object
+            Input node.
+
+        Returns
+        -------
+        root : immutable object
+            Root node of `x`.
+        """
+        if x not in self._parents:
+            # add node
+            self._sizes[x] = 1
+            self._parents[x] = x
+            self._indices[x] = len(self._indices)
```

Comment:
Curious about your thoughts on this one: As far as I can tell, two objects will have different indices if and only if they are unequal. If so, you should be able to write things like `self._indices[a] == self._indices[b]` simply as `a == b`, but maybe there are other good reasons to keep `self._indices` around?

Reply:
The equality operator throws a ValueError if  e.g. `a` is a numpy int and `b` is a tuple with multiple elements.

Running them through the dict allows me to test for equality without this problem.

Reply:
Ah, hm, what's a good example? `np.int(5) == (60, 80)` gives `False` as expected. But I see what you're saying, `__eq__` can be used for all sorts of magic; NumPy arrays can produce arrays as results, and things like ` np.ones((2, 3)) == np.ones((4, 5, 6))` will err in the future.

The bad examples I can come up with are all unhashable, though, so they can never end up in the data structure to begin with.

Except they can, if the user conjures their own type with a well-defined `__hash__` but poorly defined `__eq__`; supporting all such cases seems like it's hard in practice.

Reply:
Sorry, I should have specified that it needs to be e.g. a `np.int64`, rather than `np.int`.
If I change the type in your example I get bad behaviour:
```
>>> if np.int64(5) == (60, 80): print("hello")
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Reply:
Really?

```python
Python 3.8.3 (default, Jul  2 2020, 16:21:59)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.int64(5) == (60, 80)
array([False, False])
>>> np.__version__
'1.19.1'
```
Tested on 1.18.1 and 1.18.5 as well.

Reply:
So with that example, what I was hinting at above is that "`not in`" will fail for for `a = (60, 80)` and an `np.int64` if its hash happens to agree with `a`, even with `_indices` around:

```python
In [5]: d = DisjointSet()
In [6]: a = (60, 80)
In [7]: h = hash(a)  # Something like 1524157532169088664
In [9]: b = np.int64(h)  # Happens to satisfy hash(b) = h
In [10]: d.union(a, b)
ValueError
```

So, pretty edge-casey of course. I see how `_indices` allows you to go from not being able to work with cases where `__eq__` fails, to only having to worry about cases where `__eq__` fails but `__hash__` doesn't save the day.

This opens the door for some extremely subtle bugs, but they're edge-casey enough that it's probably not a problem at all, and the pros outweight the cons. Perhaps you even have a real life use case a la `np.int64` in mind yourself?

Reply:
Thanks for your comments here. I think we are in agreement now, so marking as resolved. Feel free to unresolve if not.

Reply:
As a complete aside, I just now realized that my example actually doesn't blow up on Python 3.7 (but it does on Python 3.8) since, to my surprise, `hash(hash(a))` is not in general the same as `hash(a)` on Python 3.7. (Not sure if that's a CPython implementation detail, or what is going on though.)